M/event dispatch

mfelsche opened the pull request:
open questions:

* how/where to do the wiring of event listeners to still be test friendly
* ...

mfelsche committed:
first draft of event dispatching

Commit:
use partialfunctions for eventlisteners

matching on event classes

Commit:
simplify eventdispatcher

mfelsche committed:
moved AccountHelper and created an interface for it

(better testing)

mfelsche committed:
add scalac postfixops

mfelsche committed:
make ApplicationContext Closeable and close it on app close

mfelsche committed:
create some events, throw them and listen to PreKeyEvents

by refreshing prekeys if we have too less

mfelsche committed:
using eventually in event dispatcher spec

as event dispatching is asynchronous

ayoub-benali commented on src/main/scala/de/m7w3/signal/ApplicationContext.scala:
```diff
@@ -1,20 +1,24 @@
 
 
-case class ContextBuilder(config: Config.SignalDesktopConfig) extends ApplicationContext{
+case class ContextBuilder(config: Config.SignalDesktopConfig,
```
why does the context builder takes an application context ?

mfelsche replied:
It only takes a reference that will be filled once the context is built.
Sounds strange, now that i write about it.
will change.

Commit:
move appCtx reference to ui parts

from ContextBuilder where it didnt belong in the first place

ayoub-benali commented on src/main/scala/de/m7w3/signal/messages/MessageReceiver.scala:
```diff
@@ -51,17 +50,16 @@ case class MessageReceiver(cipher: SignalServiceCipher,
+      if (envelope.isReceipt) {
+        logger.debug(s"received receipt from ${envelope.getSource}")
+        // not handled yet, just an example
+        eventPublisher.publishEvent(ReceiptEvent(envelope.getSourceAddress, envelope.getSourceDevice, envelope.getTimestamp))
```
tip if I may: You can create an alternative constructor for `ReceiptEvent` which takes an `Envelope` and does the fields extractions for you. So that when we need to create `ReceiptEvent` several times, we don't have to get the fields each time

mfelsche replied:
been there, done that

Commit:
simplified applicationContext initialization

and actually hook up the PreKeyRefreshListener during application context initialization
